signer/src/transaction_signer.rs

```diff
@@ -783,6 +1012,91 @@ where
         Ok(())
     }
 
+    async fn ensure_rotate_key_state_machine(
+        &mut self,
+        bitcoin_chain_tip: model::BitcoinBlockHash,
+        aggregate_key: PublicKeyXOnly,
+    ) -> Result<StateMachineId, Error> {
+        let state_machine_id = StateMachineId::RotateKey(aggregate_key, bitcoin_chain_tip);
+
+        match self.wsts_frost_state_machines.get_mut(&state_machine_id) {
+            Some(_) => {}
+            None => {
+                let storage = self.context.get_storage();
+
+                let dkg_shares = storage
+                    .get_encrypted_dkg_shares(aggregate_key)
+                    .await?
+                    .ok_or_else(|| {
+                        tracing::warn!("no DKG shares found for requested aggregate key");
+                        Error::MissingDkgShares(aggregate_key)
+                    })?;
+
+                let signing_set: BTreeSet<PublicKey> = dkg_shares
+                    .signer_set_public_keys
+                    .into_iter()
+                    .collect::<BTreeSet<_>>();
+                // This `as` cast should always be safe as our signer cap is 128.
+                let num_signers = signing_set.len() as u16;
+
+                tracing::debug!(%num_signers, "creating now frost coordinator to track pre-rotate-key validation signing round");
+                let coordinator = FrostCoordinator::load(
+                    &storage,
+                    aggregate_key,
+                    signing_set,
+                    dkg_shares.signature_share_threshold,
+                    self.signer_private_key,
+                )
+                .await?;
+
+                self.wsts_frost_state_machines
+                    .get_or_insert_mut(state_machine_id, || coordinator);
+            }
+        }
+
+        Ok(state_machine_id)
+    }
+
+    #[tracing::instrument(skip_all)]
+    async fn handle_rotate_key_message(
+        &mut self,
+        aggregate_key: PublicKeyXOnly,
+        id: StateMachineId,
+        msg: &WstsNetMessage,
+    ) -> Result<(), Error> {
+        let state_machine = self.wsts_frost_state_machines.get_mut(&id);
+        let Some(state_machine) = state_machine else {
+            tracing::warn!("missing frost coordinator for pre-rotate-key validation");
+            return Err(Error::MissingFrostStateMachine(Box::new(aggregate_key)));
+        };
+
+        tracing::info!(?msg, "processing frost coordinator message");
+
+        let (_, result) = state_machine.process_message(msg)?;
+
+        match result {
+            Some(OperationResult::SignSchnorr(_)) => {
+                tracing::info!("successfully completed pre-rotate-key validation signing round");
+                self.wsts_frost_results.put(id, true);
+                self.wsts_frost_state_machines.pop(&id);
```

Comment:
sanity check: are we sure that frost coordinator will succeed only after receiving all the participants signature shares? and not (eg) requiring all the nonces but then succeeding after receiving the first `threshold` signature shares?

(why here? I was scared by the `.pop`, but given this is the passive coordinator it doesn't matter; still, the sanity checks remain)

Reply:
@matteojug here's the code which specifies that the FROST coordinator will wait for all signers' signature shares:

https://github.com/Trust-Machines/wsts/blob/main/src/state_machine/coordinator/frost.rs#L452